src/main/java/org/deepsymmetry/beatlink/data/OpusProvider.java

```diff
@@ -158,6 +160,55 @@ public FileSystem getFileSystem() {
      */
     private final Map<Integer, LinkedBlockingQueue<MediaDetails>> archiveAttachQueueMap = new ConcurrentHashMap<>();
 
+    /**
+     * Object to store a rekordbox ID and USB slot number together.
+     */
+    public static class DeviceSqlRekordboxIdAndSlot {
+        private final int rekordboxId;
+        private final int usbSlot;
+    
+        private DeviceSqlRekordboxIdAndSlot(int rekordboxId, int usbSlot) {
+            this.rekordboxId = rekordboxId;
+            this.usbSlot = usbSlot;
+        }
+    
+        public int getRekordboxId() {
+            return rekordboxId;
+        }
+    
+        public int getUsbSlot() {
+            return usbSlot;
+        }
+    }
+
+    /**
+     * A SHA1 hash of the song structure of a track, mapped to the corresponding DeviceSQL rekordbox ID and USB slot.
+     */
+    private final Map<String, DeviceSqlRekordboxIdAndSlot> pssiToDeviceSqlRekordboxId = new ConcurrentHashMap<>();
+
+    /**
+     * Compute the SHA1 hash of the given data, similar to {@link SignatureFinder#computeTrackSignature(String, SearchableItem, int, WaveformDetail, BeatGrid)}.
+     * @param data the data to hash
+     * @return the SHA1 hash of the data, or {@code null} if there is an error
+     */
+    private String computeSha1(byte[] data) {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA1");
+            digest.update(data);
+            byte[] result = digest.digest();
+            
+            StringBuilder hex = new StringBuilder(result.length * 2);
+            for (byte b : result) {
+                hex.append(Integer.toString((b & 0xff) + 0x100, 16).substring(1));
```

Comment:
I know you got this code from me, but looking at it again, I knew there had to be a better way. Sadly, the *much* better way (the [`HexFormat` class](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/HexFormat.html)) is not available until Java 17, and we still support Java 11. But I did find a way to do at least a *little* better, and am changing `SignatureFinder` to simplify this line as follows:
```java
hex.append(String.format("%02x", (b & 0xff)));
```

Reply:
Ok, I applied this change as well (not pushed yet though).